keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,544 @@
+  proposed enhancement, is given early access to it via a fork of
+  controller-tools with the requisite generators, and is given the opportunity
+  to try it out and provide feedback.
+
+### Non-Goals
+
+Enhancements to client-go's dynamic client. The client-go dynamic client already
+supports Apply via Patch, which is adequate for the dynamic client.
+
+Protobuf support. Apply does not support protobuf, and it will not be added with
+this enhancement.
+
+## Proposal
+
+`Apply` functions should be included in the typed clients generated for
+client-go and should accept the apply configuration using a strongly typed
+representation, which will need to be generated for this purpose.
+
+### Risks and Mitigations
+
+#### Poor adoption
+
+Risk: Developers adoption is poor, either because the asthetics/ergonomics are
+not to their liking or the functinality is insufficient to do what they need to
+do with it. This could lead to (a) poor server side apply adoption, and/or (b)
+developers building alternate solutions.
+
+Mitigation: We are working with the kubebuilder community to
+get hands on feedback from developers to guide our design decisions around
+asthetics/egronomics with a goal of having both client-go and kubebuilder
+take an aligned approach to adding apply to clients in a typesafe way.
+
+## Design Details
+
+### Apply functions
+
+The client-go typed clients will be extended to include Apply functions, e.g.:
+
+```go
+func (c *deployments) Apply(ctx Context, deployment *appsv1apply.Deployment, fieldManager string, metav1.ApplyOptions) (*Deployment, error)
+```
+
+`ApplyOptions` will be added to metav1 even though `PatchOptions` will continue
+to be used over the wire. This will make it obvious in the Apply function
+signature that `fieldManager` is required.
+
+```go
+type ApplyOptions struct {
+  DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,1,rep,name=dryRun"`
+  Force *bool `json:"force,omitempty" protobuf:"varint,2,opt,name=force"`
+}
+
+func (ApplyOptions) ToPatchOptions(fieldManager string) PatchOptions
+```
+
+Apply will combine the `fieldManager` argument with `ApplyOptions` to create the
+`PatchOptions`.
+
+Each apply call will be required to provide a fieldmanager name. We will not
+provide a a way for the fieldmanager name to be set for the entire
+clientset. There are a couple reasons for this:
+
+- If a client has multiple code paths where it makes apply requests to the same
+  object, but with different field sets, they must use different field manager
+  names. If they use the same field manager name they will cause fields to be
+  accidentally removed or disowned. This is a potential foot gun we would like to
+  avoid.
+
+- Apply requests always conflict with update requests, even if they were made by
+  the same client with the same field manager name. This is by design. So when a
+  controller migrates from update to apply, it will need to deal with conflicts
+  regardless of what field manager name is used.
```

Comment:
I don't understand what this has to do with forcing people to specify a name (the first point is sufficient to carry the argument).

Reply:
You're right. Removing.